### PR TITLE
Integrate the IntRange Checker with the Checker Framework

### PR DESCRIPTION
**Pull request by JasonMrX**
Formatted and tested thoroughly. Alias for android.support.intrange is provided. Working on the manual section. Will finish the section once the code has been reviewed and accepted. 

Thanks,

Jason

### CODE REVIEW

File: checker/src/org/checkerframework/checker/intrange/IntRangeVisitor.java [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+    @Override
+    protected IntRangeAnnotatedTypeFactory createTypeFactory() {
+        return new IntRangeAnnotatedTypeFactory(checker);
+    }
+
+    /**
+     * Issues a warning if from &gt; to
+     */
+    @Override
+    public Void visitAnnotation(AnnotationTree node, Void p) {
```
Comment (wmdietl) [inline checker/src/org/checkerframework/checker/intrange/IntRangeVisitor.java]: This seems like a very low-level way of checking well-formedness. Wouldn't it be easier to write a TypeValidator subclass?

File: checker/src/org/checkerframework/checker/intrange/utils/Range.java [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+    public Range(long value) {
+        this(value, value);
+    }
+
+    public Range() {
+        this.from = Long.MIN_VALUE;
+        this.to = Long.MAX_VALUE;
+    }
+
+    private Range getRangeFromPossibleValues(long[] possibleValues) {
```
Comment (wmdietl) [inline checker/src/org/checkerframework/checker/intrange/utils/Range.java]: Should this method be static?

File: checker/src/org/checkerframework/checker/intrange/IntRangeVisitor.java [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+
+        if (!elem.toString().equals(IntRange.class.getName())) {
+            return super.visitAnnotation(node, p);
+        }
+
+        if (args.size() == 2
+                && args.get(0).getKind() == Kind.ASSIGNMENT
+                && args.get(1).getKind() == Kind.ASSIGNMENT) {
+
+            ExpressionTree expFrom = ((AssignmentTree) args.get(0)).getExpression();
```
Comment (wmdietl) [inline checker/src/org/checkerframework/checker/intrange/IntRangeVisitor.java]: How do you know that `from` comes before `to`? Could somebody write `@IntRange(to=10, from=20)` and this code wouldn't complain?

File: checker/src/org/checkerframework/checker/intrange/IntRangeTransfer.java [hunk 4]
```diff
@@ -0,0 +1,427 @@
...
+import org.checkerframework.javacutil.AnnotationUtils;
+import org.checkerframework.javacutil.TypesUtils;
+
+/**
+ * Transfer functions for IntRange Checker
+ *
+ * @author JasonMrX
+ *
+ */
+public class IntRangeTransfer extends CFTransfer {
```
Comment (wmdietl) [inline checker/src/org/checkerframework/checker/intrange/IntRangeTransfer.java]: A lot of code in this class is similar to ValueTransfer. Conceptually, wouldn't it be nice if the `@IntVal` went to an `@IntRange` once 10 concrete values are exhausted? Combining the two into once checker might allow you to reuse a lot of code and provide nicer granularity.

**Comment by wmdietl** [comment]
Mentioning @JasonMrX to ensure you get notified.

**Comment by JasonMrX** [comment]
Hi @wmdietl,

Thanks so much for the detailed response! I think the suggestion that the `@IntVal` went to an `@IntRange` once 10 concrete values are exhausted is nice and should be doable. I will think about the idea and see if I could come up with a plan to realize it. Again, thanks!

Jason

**Comment by wmdietl** [comment]
Thanks for working on these comments, @JasonMrX !
In general, I would guess that in a combined system, `@IntVal({1, 2, 10})` should be a subtype of `@IntRange(from=1, to=10)`.

**Comment by wmdietl** [comment]
I've discussed this some more: can you try merging the IntRangeChecker with the Constant Value Checker in checker-framework/framework/src/org/checkerframework/common/value/ ?
In particular, the ValueTransfer class is very similar to the IntRangeTransfer class and unifying the two hopefully simplifies the overall logic.

**Comment by JasonMrX** [comment]
Hi @wmdietl,

Yes, I will try to do that and see if we can actually merge the two. Will let you know! Again, thanks.

Jason

**Comment by JasonMrX** [comment]
Close the PL for now and will make a new one when done with the merging of the IntRange and the Constant Value

**Pull request closed without merging**